src/core/exec-invoke.c

```diff
@@ -2219,6 +2220,79 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
         return 0;
 }
 
+static int bpffs_prepare(
+                int *parent_fd,
+                PidRef *bpffs_pid) {
+
+        _cleanup_close_pair_ int bpffs_fds[2] = EBADF_PAIR;
+        int r;
+
+        assert(parent_fd);
+
+        r = socketpair(AF_UNIX, SOCK_SEQPACKET, 0, bpffs_fds);
+        if (r < 0)
+                return r;
+
+        r = pidref_safe_fork("(sd-bpffs)", FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL, bpffs_pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                _cleanup_close_ int fs_fd = -EBADF, mnt_fd = -EBADF;
+
+                fs_fd = receive_one_fd(bpffs_fds[0], 0);
+                if (fs_fd < 0)
+                        _exit(EXIT_FAILURE);
+
+                r = fsconfig(fs_fd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
+                if (r < 0)
+                        _exit(EXIT_FAILURE);
+
+                mnt_fd = fsmount(fs_fd, 0, 0);
+                if (mnt_fd < 0)
+                        _exit(EXIT_FAILURE);
+
+                r = send_one_fd(bpffs_fds[0], mnt_fd, 0);
+                if (r < 0)
+                        _exit(EXIT_FAILURE);
+
+                _exit(EXIT_SUCCESS);
+        }
+
+        *parent_fd = TAKE_FD(bpffs_fds[1]);
+
+        return 0;
+}
+
+static int bpffs_finalize(int bpffs_fd, int pid) {
```

Comment:
Missing assertions for `bpffs_fd`, also rename to `pipe_fd` or so.
Please use `PidRef` rather than `int pid`.

Also I'd like to rename `bpffs_fd` -> `pipe_fd` or so to indicate it is a socket.

Reply:
and put `TAKE_PIDREF(ret_pid)` after the call to bpffs_finalize()

Reply:
Sorry, I cannot follow?

Reply:
sorry, I commented under the wrong place